[dart 3.9.0] Hooks guide

antfitch opened the pull request:
Adds a hooks user guide to the Dart docs.

Preview:

* https://dart-dev--pr6759-dart-hooks-guide-n1bt0qgh.web.app/tools/hooks
* https://dart-dev--pr6759-dart-hooks-guide-n1bt0qgh.web.app/interop/c-interop#build-hooks

antfitch commented:
@IldySilva looping you in. Don't make comments on this yet (@dcharkes will make changes first, and then I will make revisions following that). I'll let you know when I've finished so you can take a look. I'm thinking I'll create the first version of this and merge it, then if you want to follow up and make updates, that could work.

dcharkes commented:
Hey @antfitch I've done some rewriting!

I've split the page in two toplevel headers: The first one introduces all the concepts, and the second part is the tutorial that dives deep in two of the concepts.

I haven't fixed up all the references to the packages etc.

antfitch commented:
> Hey @antfitch I've done some rewriting!
> 
> I've split the page in two toplevel headers: The first one introduces all the concepts, and the second part is the tutorial that dives deep in two of the concepts.
> 
> I haven't fixed up all the references to the packages etc.

Okay great! I want to stay away from tutorials, but definitely we need to have how-tos. I've updated the headings to reflect the structure I would like for us to use. I need you to look carefully at the link hook sections and the code that I grabbed from the example project. It looks like we're referencing an asset type that is not yet available, so that will need updating.

antfitch commented on src/content/tools/hooks.md:
```diff
@@ -0,0 +1,287 @@
+    import 'dart:convert';
+    import 'dart:io';
+
+    import 'package:data_assets/data_assets.dart';
```
I don't think we support this yet, correct?

dcharkes replied:
Correct.

(Long answer: The data_assets package can be used in the hook during builds. But
the data assets cannot yet be accessed at runtime by the Dart code.)

dcharkes commented on src/content/tools/hooks.md:
```diff
@@ -0,0 +1,377 @@
+
+## Use a hook
+
+To setup a hook, use a hook to generate assets, and use
```
What did you want to say here?

To setup a hook use a hook sounds weird.

To add assets to your project, use ...

dcharkes commented on src/content/tools/hooks.md:
```diff
@@ -0,0 +1,377 @@
+packages.
+:::
+
+Example dependencies for a build hook:
```
It doesn't make sense to use link hooks without a build hook. The link hook is
meant to make assets from a build hook smaller.

Also, both build and link hooks can work with code assets. (And once data assets
support is added both hooks can work with data assets.)

dcharkes commented on src/content/tools/hooks.md:
```diff
@@ -0,0 +1,377 @@
+
+    import 'package:data_assets/data_assets.dart';
+    import 'package:hooks/hooks.dart';
+    import 'package:record_use/record_use.dart';
```
Record use is very experimental, we shouldn't show this yet.

Note that without record use you can't really use a link hook to shrink assets
yet. So maybe it would be better to leave link hooks out completely for now.

dcharkes commented on src/content/tools/hooks.md:
```diff
@@ -0,0 +1,377 @@
+### Reference assets
+
+Assets are the files that hooks create in the
+`build/` directory. Once an asset has been created,
```
This is not true. It might be true for flutter, but not for Dart.

`dart build` creates a `build/` directory.

The assets are actually generated in `.dart_tool/build_runner/<...>/`, but I'm
not sure how much of that we want to tell users. They can see this from
inspecting `input.outputDirectoryShared`.

antfitch replied:
Cool. I've updated the docs.

dcharkes commented:
@antfitch The link hooks don't serve a use case yet before the `record_use` becomes stable. (It's currently highly experimental, not in preview yet.) The data assets are not supported either at runtime in both Dart and Flutter.

So, that leaves the features to be shipped to 3.10 stable:

* build hooks
* code assets

But I'd like to "hooks" and "assets" introduced as separate concepts so that we can add more hooks and assets later to this page. Do you think the current way the headers are structured is okay for that?

antfitch commented:
> @antfitch The link hooks don't serve a use case yet before the `record_use` becomes stable. (It's currently experimental, not in preview yet.) The data assets are not supported either at runtime in both Dart and Flutter.
> 
> So, that leaves the features to be shipped to 3.9 betas and 3.10 stable:
> 
> * build hooks
> * code assets
> 
> But I'd like to "hooks" and "assets" introduced as separate concepts so that we can add more hooks and assets later to this page. Do you think the current way the headers are structured is okay for that?
> 
> (So technically, link hooks are shipped in 3.9 betas and 3.10 stable, but they don't serve a use case yet.)

I think that sounds great. I've made some updates while preserving the existing structure so we can easily add on to it in the future. Link hooks are going to be pretty cool when all of the parts are ready.

goderbauer commented on src/content/tools/hooks.md:
```diff
@@ -0,0 +1,277 @@
+Dart, such as C, C++, Rust, or Go. `CodeAsset` is part of
+the `code_asset` package and are called with `dart:ffi` from
+the Dart code at runtime using the `@Native` annotation.
+A code asset can be used with build and link hooks.
```
Should we remove the referene to link hook here since we explain nowhere what that is yet?

goderbauer commented on src/content/tools/hooks.md:
```diff
@@ -0,0 +1,277 @@
+1.  In your Dart project, create or open `hooks/build.dart`. 
+
+1.  In the main method, add a `build` function from
+    `package:hooks/hooks.dart` and link to the native
+    library. For example:
```
instead of "and link to the native library" maybe: "... and use the appropriate toolchain to compile the native library."

goderbauer commented on src/content/tools/hooks.md:
```diff
@@ -0,0 +1,277 @@
+
+### Generate or download assets {: #generate-assets }
+
+Assets are generated and/or downloaded when you invoke your
```
Since you never really invoke a hook directly, maybe better: Assets are generated and/or downloaded according to your hook script when your hook is invoked by the Dart SDK, which autoamtically happens when you use the `run`, `build`, or `test` command on your package.

goderbauer commented on src/content/tools/hooks.md:
````diff
@@ -0,0 +1,277 @@
+      });
+    }
+    ```
+
````
Maybe this could get an introductory sentence like: Build hooks are given two arguments:

goderbauer commented on src/content/tools/hooks.md:
````diff
@@ -0,0 +1,277 @@
+    }
+    ```
+
+    * `input`: The input for `hook/build.dart`. This
````
Maybe it would be useful to include that the `input` is read-only and it includes information for the hook to produce the right asset type, e.g. like target OS etc.

antfitch replied:
Yes, definitely. Updated!

goderbauer commented on src/content/tools/hooks.md:
```diff
@@ -0,0 +1,277 @@
+      output directory, etc. For details, see
+      the [`BuildInput`][] class.
+
+    * `output`: The output for `hook/build.dart`. This
```
Maybe include here that this is a builder object and it is the responsibility of the build hook to set the appropriate information on it. What asset have been produced? etc.

In other words (and maybe we can include this somewhere in the intro sentence above to connect it all together): The build hook script is supposed to read from the inputs, produce an asset, and then fill the output with what it produced.

antfitch replied:
Thanks for the clarification. Please take a close look at my changes.

goderbauer commented on src/content/tools/hooks.md:
```diff
@@ -0,0 +1,277 @@
+
+Assets are the files that hooks create. Once an asset has been created,
+you can reference it in your code and at runtime with it's
+asset ID ([`assetId`][]).
```
I feel like there is kind of a disconnect between this sentence and the example below because the example below never refers to an asset id.

dcharkes replied:
I've reworked it.

I kept all the asset-id stuff in the reference-asset section, because you don't
really think about it during the build-time. The alternative would be to
already explain it in the create-hook section, saying that it will be important
later. WDYT?

goderbauer replied:
I think this is great now. Thanks for the rewrite!

goderbauer commented on src/content/tools/hooks.md:
```diff
@@ -0,0 +1,277 @@
+Dart hooks. Your Dart hooks are invoked when you use the
+`run`, `build`, or `test` command on your package.
+The resulting assets are stored in the directory that you
+specified with [`BuildInput.sharedOutputDirectory`][]. 
```
I am not sure what this sentence is supposed to tell me. It sounds like it wants to tell me how to set the output directory for an asset, but it then links to`BuildInput.sharedOutputDirectory`, which is a getter.

There's also kind of a disconnect to the example given above, which doesn't seem to specify an output directory at all (it probably happens indirectly inside the CBuilder). Can we somehow connect this back to the example so people are not confused?

goderbauer reviewed (commented):
Thank you for writing this up @antfitch 🚀

antfitch commented on src/content/tools/hooks.md:
```diff
@@ -0,0 +1,277 @@
+
+### Build hooks {:.no_toc}
+
+A build hook builds assets that you want
+to run during your project's build process. This is useful
```
Yes, this is confusing. I've updated this section in my next revision. Please take a close look at my changes.

antfitch commented on src/content/tools/hooks.md:
```diff
@@ -0,0 +1,277 @@
+  runtime.
+
+A package's build hook is placed in the `hook/build.dart`
+directory. Use the [`build`][] function to parse the hook
```
Added and made some updates in this section.

antfitch commented on src/content/tools/hooks.md:
```diff
@@ -0,0 +1,277 @@
+
+### Test assets {: #test-assets }
+
+After you've generated an asset and referenced it in
+your Dart code, consider creating a test for the generated
+asset. In the following example, a test is created for
```
Sounds good. Added updates.

antfitch commented on src/content/tools/hooks.md:
```diff
@@ -0,0 +1,277 @@
+your Dart code, consider creating a test for the generated
+asset. In the following example, a test is created for
+`native_add_library.dart`, a script that references a
+code asset called  
```
Fixed!

antfitch commented:
@goderbauer and @dcharkes, please take another look. I've closed any requests that I've completed (clutter cleanup) and left open the ones that need help. Feel free to open any closed requests if an issue persists.

dcharkes commented on src/content/tools/hooks.md:
```diff
@@ -0,0 +1,293 @@
+The resulting assets are stored in the directory that you
+specified in your hooks script. 
+
+### Reference assets {: #reference-assets }
```
Maybe this should be "Use assets"

Basically the build step happens at "build time" and using an asset happens at
"run time". Would standardizing on build-time and runtime in this doc clarify
things?

(Using "create" as the verb for assets at build-time is fine. Create and use.
Create and reference.)

goderbauer replied:
```suggestion
### Reference assets from Dart code {: #reference-assets }
```

antfitch replied:
I think `Use assets` will work the best since it's brief, looks good in the TOC, and is easy to parse quickly. @dcharkes and @goderbauer fyi.

goderbauer reviewed (commented):
LGTM

This is great! Thank you!

antfitch commented:
> LGTM

Thank you!!!!

parlough commented on src/content/tools/hooks.md:
```diff
@@ -0,0 +1,294 @@
+version: 0.1.0
+
+environment:
+  sdk: '>=3.9.0 <4.0.0'
```
```suggestion
  sdk: ^3.9.0
```

parlough commented on src/content/tools/hooks.md:
```diff
@@ -0,0 +1,294 @@
+  sdk: '>=3.9.0 <4.0.0'
+
+dependencies:
+  ... 
```
So that this file can be valid yaml, in case it's ever copy pasted.

```suggestion
  # ... 
```

parlough commented on src/content/tools/hooks.md:
```diff
@@ -0,0 +1,294 @@
+  native_toolchain_c: any
+
+dev_dependencies:
+  ...
```
```suggestion
  # ...
```

parlough commented on src/content/tools/hooks.md:
```diff
@@ -0,0 +1,294 @@
+Afterwards, these assets can be called from the Dart code of
+a package.
+
+A package's build hook is automatically invoked by the
```
Can a package only have one build hook?

parlough commented on src/content/tools/hooks.md:
```diff
@@ -0,0 +1,294 @@
+Assets are the files that hooks create. Once an asset is
+created, you can reference it in your code and at runtime
+with its asset ID ([`assetId`][]). Asset IDs are structured
+as `package:<package-name>/<asset-name>`. Build hooks may
```
https://developers.google.com/style/word-list#may

```suggestion
as `package:<package-name>/<asset-name>`. Build hooks can
```

parlough commented on src/content/tools/hooks.md:
```diff
@@ -0,0 +1,294 @@
+### Use assets {: #reference-assets }
+
+Assets are the files that hooks create. Once an asset is
+created, you can reference it in your code and at runtime
```
```suggestion
created, you can reference it in your code and at run time
```

parlough commented on src/content/tools/hooks.md:
```diff
@@ -0,0 +1,294 @@
+library URI. In the previous example, this is
+`package:native_add_library/native_add_library.dart`, which
+is the same asset ID as output from the build hook. This
+enables Dart to connect an asset referenced at runtime to
```
```suggestion
enables Dart to connect an asset referenced at run time to
```

parlough commented on src/content/tools/hooks.md:
```diff
@@ -0,0 +1,294 @@
+
+### Add dependencies {: #add-dependencies-hooks-assets }
+
+To use a hook, you must first add it to your `pubspec.yaml`
+dependencies.
```
Unless there's a specific reason developers need to add these dependencies to the file manually, do suggest they use `dart pub add hooks code_assets` instead. It's what we've been moving to elsewhere and is less prone to mistakes.

parlough commented on src/content/tools/hooks.md:
```diff
@@ -0,0 +1,294 @@
+The assets produced for your package might depend on
+[`assets`][] or [`metadata`][] produced by the build hooks
+from the packages in your direct dependencies. Therefore,
+build hooks are run in the order of dependencies.
```
The order of what dependencies? From a package's pubspec file, declared in the hook output?

dcharkes replied:
Good catch! The pubspec dependencies.

parlough commented on src/content/tools/hooks.md:
```diff
@@ -0,0 +1,294 @@
+
+1.  In your Dart project, create or open `hooks/build.dart`. 
+
+1.  In the `main` method, add a `build` function from
+    `package:hooks/hooks.dart` and use the appropriate
```
```suggestion
1.  In the `main` method, call the `build` function from
    `package:hooks/hooks.dart` and use the appropriate
```

parlough commented on src/content/tools/hooks.md:
```diff
@@ -0,0 +1,294 @@
+
+    The `build` function takes two arguments:
+
+    * `input`: (readonly) The input for the hook.
```
We don't use this `(readonly)` syntax anywhere else, so consider including that information directly in the sentence.

```suggestion
    * `input`: The read-only input for the hook.
```

parlough commented on src/content/tools/hooks.md:
```diff
@@ -0,0 +1,294 @@
+      target OS, target architecture, output directory, and
+      more). For details, see the [`BuildInput`][] class.
+
+    * `output`: (writeonly) The output for the hook. After the
```
```suggestion
    * `output`: The write-only builder for the hook output. After the
```

parlough commented on src/content/tools/hooks.md:
```diff
@@ -0,0 +1,294 @@
+
+### Automatically add assets {: #generate-assets }
+
+Assets are added indirectly when you use the `run`,
+`build`, or `test` command to invoke hooks on your package.
```
```suggestion
Assets are added indirectly when you use the `run`,
`build`, or `test` commands to invoke hooks on your package.
```

parlough commented on src/content/tools/hooks.md:
```diff
@@ -0,0 +1,294 @@
+    void main(List<String> args) async {
+      await build(args, (input, output) async {
+        final packageName = input.packageName;
+        final cbuilder = CBuilder.library(
+          name: packageName,
+          assetName: '$packageName.dart',
+          sources: ['src/$packageName.c'],
+        );
+        await cbuilder.run(
```
```suggestion
        final cBuilder = CBuilder.library(
          name: packageName,
          assetName: '$packageName.dart',
          sources: ['src/$packageName.c'],
        );
        await cBuilder.run(
```

parlough commented on src/content/tools/hooks.md:
```diff
@@ -0,0 +1,294 @@
+`package:native_add_library/native_add_library.dart`, and is
+based on the `packageName` and `assetName`.
+
+The following example illustrates how to call the native
+C function `add` from `native_add_library.c`:
```
It seems the following code snippet is not necessarily calling the function, just initially binding to it. Perhaps change this to "how to bind" or add an example of calling it as well.

parlough commented on src/content/tools/hooks.md:
```diff
@@ -0,0 +1,294 @@
+
+[Hooks package]: {{site.pub-pkg}}/hooks
+[Hooks library reference]: {{site.pub-api}}/hooks/latest/hooks/
+[Build and bundle native code]: https://dart.dev/interop/c-interop#build-hooks
```
```suggestion
[Build and bundle native code]: /interop/c-interop#build-hooks
```

parlough reviewed (approved):
Overall this looks amazing to me. I left a some comments and a few questions, but if you'd prefer to defer addressing any of them, that's fine as well. Thanks everyone!

goderbauer commented on src/content/tools/hooks.md:
```diff
@@ -0,0 +1,300 @@
+📁  example_project                   // Project with hooks.
+    📁  hook                          // Add hook scripts here.
+        📄 build.dart
+    📁  lib                           // Us your assets here.
```
```suggestion
    📁  lib                           // Use your assets here.
```

goderbauer reviewed (approved):
Thank you everyone!

antfitch commented on src/content/tools/hooks.md:
```diff
@@ -0,0 +1,294 @@
+### Build hooks {:.no_toc}
+
+With build hooks, a package can do things such as
+compile or download native assets such as C or Rust libraries.
```
Added the entry to this list for now: https://github.com/dart-lang/site-www/issues/6461
